lib/pbench/server/cache_manager.py

```diff
@@ -194,6 +181,82 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class LockRef:
+    """Keep track of a cache lock passed off to a caller"""
+
+    def __init__(self, lock: Path, exclusive: bool = False, wait: bool = True):
+        """Initialize a lock reference
+
+        The lock file is opened in "w+" mode, which is "write update": unlike
+        "r+", this creates the file if it doesn't already exist, but still
+        allows lock conversion between LOCK_EX and LOCK_SH.
+
+        Args:
+            lock: the path of a lock file
+            exclusive: lock for exclusive access
+            wait: wait for lock
+        """
+        self.lock = lock.open("w+")
+        self.locked = False
+        self.exclusive = exclusive
+        self.unlock = True
+        self.wait = wait
+
+    def acquire(self) -> "LockRef":
+        """Acquire the lock"""
+
+        cmd = fcntl.LOCK_EX if self.exclusive else fcntl.LOCK_SH
+        if not self.wait:
+            cmd |= fcntl.LOCK_NB
+        fcntl.lockf(self.lock, cmd)
+        self.locked = True
+        return self
+
+    def release(self):
+        """Release the lock and close the lock file"""
+
+        exception = None
+        try:
+            fcntl.lockf(self.lock, fcntl.LOCK_UN)
+            self.lock.close()
+            self.locked = False
+            self.exclusive = False
+        except Exception as e:
+            exception = e
+        finally:
+            # Release our reference to the lock file so that the
+            # object can be reclaimed.
+            self.lock = None
+        if exception:
+            raise exception
+
+    def upgrade(self):
+        """Upgrade a shared lock to exclusive"""
+        if not self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_EX)
+            self.exclusive = True
+
+    def downgrade(self):
+        """Downgrade an exclusive lock to shared"""
+        if self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_SH)
+            self.exclusive = False
+
+    def keep(self) -> "LockRef":
+        """Tell the context manager not to unlock on exit"""
+        self.unlock = False
+        return self
```

Comment:
This seems like it would be better as a creation attribute than as a "setter".  (The "fluent" interface is cute, but if this were set when the object was instantiated, then we wouldn't need the fluency.)

Reply:
Actually, no; if we fail out before calling the setter, the context manager cleans up after itself without extra logic.

Reply:
Hmmm...interesting....  That feature is not obvious...it would be good if that were highlighted in a code comment.